daphne_worker/src/durable/mod.rs

```diff
@@ -209,97 +190,185 @@ impl<'srv> DurableConnector<'srv> {
+                    };
                 }
-                *self.touched() = true;
+
+                let span = create_span_from_request(&req);
+                <$name as GcDurableObject>::handle(self, req).instrument(span).await
+            }
+
+            async fn alarm(&mut self) -> Result<Response> {
+                self.state.storage().delete_all().await?;
+                ::tracing::trace!(
+                    "{}: deleted instance {}",
+                    ::std::stringify!($name),
+                    self.state.id().to_string()
+                );
+                ::worker::Response::from_json(&())
             }
-            _ => {}
         }
-        Ok(std::ops::ControlFlow::Continue(req))
+    };
+}
+
+/// Ensure that an alarm for self deletion is setup for a durable object.
+async fn ensure_alarmed(
+    state: &State,
+    lifetime: ScheduledTime,
+    deployment: DaphneWorkerDeployment,
+) -> Result<()> {
+    if let Err(e) = state.storage().set_alarm(lifetime).await {
+        if matches!(deployment, DaphneWorkerDeployment::Dev) {
+            warn!("ignoring get_alarm() failure in a dev environment until --experimental-local implements it: {e}");
+        } else {
+            // We only return an error if not in the "dev" deployment as
+            // the experimental-local dev environment doesn't have
+            // working get_alarm() and set_alarm() yet, so we want to
+            // ignore errors in that case.
+            return Err(e);
+        }
+    }
+    Ok(())
+}
+
+/// Register a do to be deleted by the [`GarbageCollector`](garbage_collector::GarbageCollector).
+///
+/// If, however, the request is a [`GarbageCollector::DeleteAll`] request we also handle that
+/// request. This consumes the request and returns `Continue::Break`, signaling that the durable
+/// object should itself return from fetch.
+///
+/// If the request *wasn't* a [`GarbageCollector::DeleteAll`] request, the `Request` is
+/// returned through the `ControlFlow::Continue` variant.
+///
+/// # Note
+/// This function is doing two things at once, which makes it confusing, but both things revolve
+/// around keeping some state (`gc_delete_all_is_setup`) in sync and such it's less error prone for
+/// callers if it's all done in one place.
+async fn setup_and_handle_garbage_collector_requests<T: GcDurableObject>(
+    state: &State,
+    env: &Env,
+    req: Request,
+) -> Result<ControlFlow<(), Request>> {
+    // We first check if this was a delete all request. If it was then it means that the setup in
+    // the else branch was already done before.
+    if let Some(TestStateCleaner::DeleteAll) = TestStateCleaner::try_from_uri(&req.path()) {
+        state.storage().delete_all().await?;
+        state
+            .storage()
+            .put("gc_delete_all_is_setup", &false)
+            .await?;
+        Ok(ControlFlow::Break(()))
+    } else {
+        // if this wasn't a delete all request then we must ensure that we have registered
```

Comment:
Don't we only want to do this in testing?
```suggestion
        // If this wasn't a delete all request then we must ensure that we have registered
```

Reply:
We check that before calling the function at all, you can see that in the macro. In production we don't even call this.